File: enigma/Reflector.py
class Reflector:
    def __init__(self, ukw_type):
        self.type = ukw_type
        self.alphabet = "abcdefghijklmnopqrstuvwxyz"
    
    def reflect(self, letter, zusStep):
        # 1. get the position of the zusatzwalze - the position of the letter in the alphabet
        # 2. Make sure it is positive an < 26
        # 3. return the letter at that position in the reflector
        return self.type[(26 - (zusStep - self.alphabet.index(letter))) % 26]

File: src/reflector.rs
pub struct Reflector {
    alphabet: String,
    sequence: String,
}

impl Reflector {
    pub fn new(sequence: String) -> Self {
        Reflector {
            alphabet: "abcdefghijklmnopqrstuvwxyz".to_string(),
            sequence,
        }
    }

    pub fn reflect(&self, letter: char, zus_step: i16) -> char {
        if let Some(position) = self.alphabet.find(letter) {
            let chars: Vec<char> = self.sequence.chars().collect();
            chars[((26 - (zus_step - (position as i16))) % 26) as usize]
        } else {
            panic!("Character {:?} not in alphabet", letter)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn should_reflect() {
        let r = Reflector::new("enkqauywjicopblmdxzvfthrgs".to_string());
        let output = r.reflect('a', 2);
        assert_eq!(output, 'g')
    }

    #[test]
    #[should_panic]
    pub fn should_panic() {
        let r = Reflector::new("enkqauywjicopblmdxzvfthrgs".to_string());
        r.reflect('ù', 2);
    }
}
